render-wasm/src/render.rs

```diff
@@ -337,58 +381,72 @@ impl RenderState {
+    ) -> Result<(), String> {
+        if !self.render_in_progress {
+            return Ok(());
+        }
 
-            let mut paint = skia::Paint::default();
-            paint.set_blend_mode(element.blend_mode().into());
-            paint.set_alpha_f(element.opacity());
-            let filter = element.image_filter(self.viewbox.zoom * self.options.dpr());
-            if let Some(image_filter) = filter {
-                paint.set_image_filter(image_filter);
-            }
+        while let Some((node_id, visited_children)) = self.pending_stack.pop() {
+            if let Some(element) = tree.get(&node_id) {
+                if !visited_children {
+                    if !node_id.is_nil() {
+                        if !element.bounds().intersects(self.viewbox.area) || element.hidden() {
+                            debug::render_debug_element(self, element, false);
+                            continue;
+                        } else {
+                            debug::render_debug_element(self, element, true);
+                        }
+                    }
 
-            let layer_rec = skia::canvas::SaveLayerRec::default().paint(&paint);
-            // This is needed so the next non-children shape does not carry this shape's transform
-            self.final_surface.canvas().save_layer(&layer_rec);
+                    let mut paint = skia::Paint::default();
+                    paint.set_blend_mode(element.blend_mode().into());
+                    paint.set_alpha_f(element.opacity());
 
-            self.drawing_surface.canvas().save();
-            if !root_id.is_nil() {
-                self.render_shape(&mut element.clone(), element.clip());
-            } else {
-                self.apply_drawing_to_final_canvas();
-            }
+                    if let Some(image_filter) =
+                        element.image_filter(self.viewbox.zoom * self.options.dpr())
+                    {
+                        paint.set_image_filter(image_filter);
+                    }
 
-            self.drawing_surface.canvas().restore();
+                    let layer_rec = skia::canvas::SaveLayerRec::default().paint(&paint);
+                    self.render_surface.canvas().save_layer(&layer_rec);
 
-            // draw all the children shapes
-            if element.is_recursive() {
-                for id in element.children_ids() {
                     self.drawing_surface.canvas().save();
-                    is_complete = self.render_shape_tree(&id, tree) && is_complete;
+                    if !node_id.is_nil() {
+                        self.render_shape(&mut element.clone(), element.clip());
+                    } else {
+                        self.apply_drawing_to_render_canvas();
+                    }
                     self.drawing_surface.canvas().restore();
+
+                    // Set the node as visited before processing children
+                    self.pending_stack.push((node_id, true));
```

Comment:
Let's rename `pending stack` to `pending_nodes`. When we were commenting the code earlier, I though this `pending_stack` contained the IDs of requestAnimationFrames (probably because we were talking about it, but still…)

Reply:
Great, thanks!